### PR TITLE
Retryable futures

### PR DESCRIPTION
**Pull request by iBlackShadow**

### CODE REVIEW

File: server/app/com/xsn/explorer/util/RetryableFuture.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  def apply[A](
+      delays: List[FiniteDuration]
+  )(
+      shouldRetry: Try[A] => Boolean
+  )(f: => Future[A])(implicit ec: ExecutionContext, scheduler: Scheduler): Future[A] = {
+    delays.foldLeft(f) {
+      case (result, delay) =>
+        result.transformWith { t =>
+          if (shouldRetry(t)) {
+            after(delay, scheduler, ec, Future.successful(1)).flatMap(_ => f)
```
Comment (AlexITC) [inline server/app/com/xsn/explorer/util/RetryableFuture.scala:29]: Another way is `scheduler.scheduleOnce(delay)(f)`, it looks simpler to me.

File: server/app/com/xsn/explorer/util/RetryableFuture.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      maxRetries: Int
+  )(implicit ec: ExecutionContext, scheduler: Scheduler) = {
+    val initialDelayMillis = initialDelay.toMillis
+    val delays = List.tabulate(maxRetries)(n => getDelay(initialDelayMillis, n))
+
+    RetryableFuture[A](delays) _
+  }
+
+  private def getDelay(baseDelay: Long, retry: Int, factor: Int = 2): FiniteDuration = {
+    (Math.pow(factor, retry) * baseDelay).longValue.millis
```
Comment (AlexITC) [inline server/app/com/xsn/explorer/util/RetryableFuture.scala]: Add a small random jitter when generating delays, `Random.nextInt(100)` should be enough.

File: server/app/com/xsn/explorer/util/RetryableFuture.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+import akka.actor.Scheduler
+import akka.pattern.Patterns.after
+
+import scala.concurrent.duration._
+import scala.util.Try
+
+object RetryableFuture {
+
+  def apply[A](
```
Comment (AlexITC) [inline server/app/com/xsn/explorer/util/RetryableFuture.scala:20]: Add documentation to each method.

File: server/test/com/xsn/explorer/util/RetryableFutureSpec.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  object UnstableOperation {
+
+    def apply(numberOfFailures: Int, error: Throwable): UnstableOperation = {
+      new UnstableOperation(numberOfFailures, error)
+    }
+  }
+
+  implicit val scheduler: Scheduler = ActorSystem().scheduler
```
Comment (AlexITC) [inline server/test/com/xsn/explorer/util/RetryableFutureSpec.scala]: You need to make sure the actor system is terminated after the test completes, for example:
```
class RetryableFutureSpec
    extends TestKit(ActorSystem("RetryableFutureSpec"))
    with WordSpecLike
    with BeforeAndAfterAll {

  override def afterAll: Unit = {
    TestKit.shutdownActorSystem(system)
  }
}
```

File: server/test/com/xsn/explorer/util/RetryableFutureSpec.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+  private val retryOnRuntimeException = (result: Try[Int]) => {
+    result.map(_ => false).recover { case _: RuntimeException => true }.getOrElse(false)
```
Comment (AlexITC) [inline server/test/com/xsn/explorer/util/RetryableFutureSpec.scala]:
```
result match {
  case Failure(_: RuntimeException) => true
  case _ => false
}
```

File: server/app/com/xsn/explorer/util/RetryableFuture.scala (hunk 1 above)
Comment (iBlackShadow) [reply server/app/com/xsn/explorer/util/RetryableFuture.scala:29]: `scheduler.scheduleOnce` returns `Cancellable` while `after` returns a `Future` so i could not make this change

Comment (AlexITC) [reply server/app/com/xsn/explorer/util/RetryableFuture.scala:29]: Oh, nice catch.

File: server/app/com/xsn/explorer/util/RetryableFuture.scala [hunk 6]
```diff
@@ -53,7 +71,16 @@ object RetryableFuture {
...
+  /**
+   * calculates the delay for the given retry number
+   * @param baseDelay the delay for the first retry
+   * @param retry the retry for which the delay is being calculated
+   * @param factor how much the delay will grow with each retry. for example:
+   *               2 -> double the delay with each retry
+   *               3 -> triples the delay with each retry
+   * @return
+   */
   private def getDelay(baseDelay: Long, retry: Int, factor: Int = 2): FiniteDuration = {
```
Comment (AlexITC) [inline server/app/com/xsn/explorer/util/RetryableFuture.scala]: Update this to `private[util] def getDelay(...)` and then, write tests for the function.

**Review by AlexITC — APPROVED** [review]

**Pull request merged**